### ISSUE TITLE
Validating a User's Token

### ISSUE THREAD

**Issue by RaginBajin** [rackspace/gophercloud#424]
There should be a way to validate a user's token.  There may be a way, but I haven't figured out how, so maybe some help with that would be great. 

The goal would be to be able to create an GO based API Server just like a Neutron Server, Nova Server, Swift API server, etc.

**Comment by jrperritt** [comment]
Currently, there are 2 ways you can handle token validation:

1. Let Gophercloud handle it for you via [automatic re-auth](https://github.com/rackspace/gophercloud/blob/master/provider_client.go#L186). You can enable this by [setting the `AllowReauth` option to `true`.](https://github.com/rackspace/gophercloud/blob/master/auth_options.go#L45)
2. Manually compare the [token's `ExpiresAt` field](https://github.com/rackspace/gophercloud/blob/master/openstack/identity/v2/tokens/results.go#L22) with the current time.

Would either of those suit your needs?

**Comment by RaginBajin** [comment]
Kinda. 

I guess I was looking for something like:  http://docs.openstack.org/developer/keystone/api_curl_examples.html#get-tokens-token-id

Where with admin credentials I can authenticate another users token to not only see if they are authorized, but it pulls back data about the user as well.

**Comment by jrperritt** [comment]
Yeah, that API route isn't currently implemented, but PRs are always welcome  :smile:

**Comment by einarf** [comment]
Identity v3 do have `Get` and `Validate`, but maybe this was for identity v2?

A major improvement would be to add a way to authenticate just with a token. This way you can avoid constantly asking for new tokens when this is not needed. Getting token information (at least in v3) returns all the information you need about the token including the service catalog (Called `AuthInfo` in the python keystone client). It is then trivial to create a `ProviderClient` directly from this info alone. This is very useful when making API servers since you can memcache the `AuthInfo` structure and convert it into a client quickly, then do operations on their behalf.

The python keystoneclient got these features through a new auth plugin when sessions were introduced. `AuthInfo` = `ProviderClient`.

**Comment by zyren88** [comment]
I just found a little of api function such as `Get` and `Validate`, but not the function to validate for signed Token local when I used PKI Token.
https://github.com/rackspace/gophercloud/blob/master/openstack/identity/v3/tokens/requests.go

are there anybody already had implimented a validate by pub key like keystoneclient(https://github.com/openstack/python-keystoneclient/blob/master/keystoneclient/middleware/auth_token.py) ?
